File: mage/src/main/java/mil/nga/giat/mage/map/preference/MapPreferencesActivity.java
package mil.nga.giat.mage.map.preference;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.preference.PreferenceCategory;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.PreferenceScreen;
import androidx.preference.SwitchPreferenceCompat;

import com.google.common.collect.Collections2;

import java.util.Collection;
import java.util.List;
import java.util.Set;

import javax.inject.Inject;

import dagger.hilt.android.AndroidEntryPoint;
import mil.nga.giat.mage.R;
import mil.nga.giat.mage.data.feed.Feed;
import mil.nga.giat.mage.sdk.datastore.layer.Layer;
import mil.nga.giat.mage.sdk.datastore.layer.LayerHelper;
import mil.nga.giat.mage.sdk.datastore.user.Event;
import mil.nga.giat.mage.sdk.datastore.user.EventHelper;

/**
 * Provides map configuration driven settings that are available to the user.
 * Check mappreferences.xml for the configuration.
 *
 * @author newmanw
 */
@AndroidEntryPoint
public class MapPreferencesActivity extends AppCompatActivity {

	public static String LOG_NAME = MapPreferencesActivity.class.getName();

	public static final int TILE_OVERLAY_ACTIVITY = 100;
	public static final int ONLINE_LAYERS_OVERLAY_ACTIVITY = 200;

	private MapPreferenceFragment preference = new MapPreferenceFragment();

	@AndroidEntryPoint
	public static class MapPreferenceFragment extends PreferenceFragmentCompat {
		private MapPreferencesViewModel viewModel;

		@Inject
		protected SharedPreferences preferences;

		@Inject
		protected MapLayerPreferences mapLayerPreferences;

		private Event event;

		@Override
		public void onCreate(@Nullable Bundle savedInstanceState) {
			super.onCreate(savedInstanceState);

			viewModel = new ViewModelProvider(this).get(MapPreferencesViewModel.class);
			viewModel.getFeeds().observe(this, this::onFeeds);
		}

		@Override
		public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
			addPreferencesFromResource(R.xml.mappreferences);
		}

		@Override
		public void onResume() {
			super.onResume();

			event = EventHelper.getInstance(getActivity().getApplicationContext()).getCurrentEvent();
			viewModel.setEvent(event.getRemoteId());

			findPreference(getString(R.string.tileOverlaysKey)).setOnPreferenceClickListener(preference -> {
				Intent intent = new Intent(getActivity(), TileOverlayPreferenceActivity.class);
				getActivity().startActivityForResult(intent, TILE_OVERLAY_ACTIVITY);
				return true;
			});

			findPreference(getString(R.string.onlineLayersKey)).setOnPreferenceClickListener(preference -> {
				Intent intent = new Intent(getActivity(), OnlineLayersPreferenceActivity.class);
				getActivity().startActivityForResult(intent, ONLINE_LAYERS_OVERLAY_ACTIVITY);
				return true;
			});

			// TODO : Remove the below and rework OverlayPreference to have a 'entities' similar to a list preference, these would be the 'display values'
			try {
				List<Layer> layers = LayerHelper.getInstance(getContext()).readByEvent(event, "GeoPackage");
				layers.addAll(LayerHelper.getInstance(getContext()).readByEvent(event, "Feature"));
				layers.addAll(LayerHelper.getInstance(getContext()).readByEvent(event, "Imagery"));
				Collection<Layer> available = Collections2.filter(layers, layer -> !layer.isLoaded());

				OverlayPreference p = (OverlayPreference) findPreference(getResources().getString(R.string.tileOverlaysKey));
				p.setAvailableDownloads(!available.isEmpty());
			} catch (Exception e) {
				Log.e(LOG_NAME, "Problem setting preference.", e);
			}
		}

		@Override
		public void onPause() {
			super.onPause();

			findPreference(getString(R.string.tileOverlaysKey)).setOnPreferenceClickListener(null);
			findPreference(getString(R.string.onlineLayersKey)).setOnPreferenceClickListener(null);
		}

		private void onFeeds(List<Feed> feeds) {
			PreferenceScreen screen = getPreferenceScreen();
			Set<String> enabledFeeds = mapLayerPreferences.getEnabledFeeds(event.getId());

			for (Feed feed : feeds) {
				PreferenceCategory category = (PreferenceCategory) screen.getPreference(screen.getPreferenceCount() - 1);
				SwitchPreferenceCompat feedPreference = mapLayerPreferences.mapFeedPreference(feed, getActivity(), enabledFeeds.contains(feed.getId()));
				feedPreference.setOnPreferenceClickListener( preference -> {
					onFeedClick(feed, feedPreference.isChecked());
					return true;
				});
				category.addPreference(feedPreference);
			}
		}

		private void onFeedClick(Feed feed, boolean on) {
			Set<String> feeds = mapLayerPreferences.getEnabledFeeds(event.getId());
			if (on) {
				feeds.add(feed.getId());
			} else {
				feeds.remove(feed.getId());
			}

			mapLayerPreferences.setEnabledFeeds(event.getId(), feeds);
		}
	}

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		getSupportFragmentManager().beginTransaction().replace(android.R.id.content, preference).commit();
	}
}

File: mage/src/main/java/mil/nga/giat/mage/map/preference/MapPreferencesActivity.kt
package mil.nga.giat.mage.map.preference

import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.lifecycle.ViewModelProvider
import androidx.preference.Preference
import androidx.preference.PreferenceCategory
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.SwitchPreferenceCompat
import dagger.hilt.android.AndroidEntryPoint
import mil.nga.giat.mage.R
import mil.nga.giat.mage.data.datasource.event.EventLocalDataSource
import mil.nga.giat.mage.data.datasource.layer.LayerLocalDataSource
import mil.nga.giat.mage.database.model.event.Event
import mil.nga.giat.mage.database.model.feed.Feed
import mil.nga.giat.mage.database.model.layer.Layer
import javax.inject.Inject

@AndroidEntryPoint
class MapPreferencesActivity : AppCompatActivity() {
   private val preference = MapPreferenceFragment()

   @AndroidEntryPoint
   class MapPreferenceFragment : PreferenceFragmentCompat() {

		@Inject lateinit var preferences: SharedPreferences
		@Inject lateinit var mapLayerPreferences: MapLayerPreferences
      @Inject lateinit var layerLocalDataSource: LayerLocalDataSource
		@Inject lateinit var eventLocalDataSource: EventLocalDataSource

      private var event: Event? = null
      private lateinit var viewModel: MapPreferencesViewModel

      override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         viewModel = ViewModelProvider(this).get(MapPreferencesViewModel::class.java)
         viewModel.feeds.observe(this) { feeds: List<Feed> -> onFeeds(feeds) }

      }

      override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
         addPreferencesFromResource(R.xml.mappreferences)
      }

      override fun onResume() {
         super.onResume()
         event = eventLocalDataSource.currentEvent
         viewModel.setEvent(event?.remoteId)

         findPreference<Preference>(getString(R.string.tileOverlaysKey))?.onPreferenceClickListener =
            Preference.OnPreferenceClickListener {
               val intent = Intent(activity, TileOverlayPreferenceActivity::class.java)
               requireActivity().startActivityForResult(intent, TILE_OVERLAY_ACTIVITY)
               true
            }

         findPreference<Preference>(getString(R.string.onlineLayersKey))?.onPreferenceClickListener =
            Preference.OnPreferenceClickListener {
               val intent = Intent(activity, OnlineLayersPreferenceActivity::class.java)
               requireActivity().startActivityForResult(intent, ONLINE_LAYERS_OVERLAY_ACTIVITY)
               true
            }
         val showMGRS = findPreference<SwitchPreferenceCompat>(getString(R.string.showMGRSKey))
         val showGARS = findPreference<SwitchPreferenceCompat>(getString(R.string.showGARSKey))
         showMGRS?.onPreferenceChangeListener =
            Preference.OnPreferenceChangeListener { _: Preference?, newValue: Any ->
               if (newValue as Boolean) {
                  showGARS?.isChecked = false
               }
               true
            }
         showGARS?.onPreferenceChangeListener =
            Preference.OnPreferenceChangeListener { _: Preference?, newValue: Any ->
               if (newValue as Boolean) {
                  showMGRS?.isChecked = false
               }
               true
            }

         // TODO : Remove the below and rework OverlayPreference to have a 'entities' similar to a list preference, these would be the 'display values'
         val availableDownLoads = mutableListOf<Layer>().apply {
            addAll(layerLocalDataSource.readByEvent(event, "GeoPackage"))
            addAll(layerLocalDataSource.readByEvent(event, "Feature"))
            addAll(layerLocalDataSource.readByEvent(event, "Imagery"))
         }.any { !it.isLoaded }

         val overlayPreference = findPreference<Preference>(resources.getString(R.string.tileOverlaysKey)) as OverlayPreference?
         overlayPreference?.setAvailableDownloads(availableDownLoads)
      }

      override fun onPause() {
         super.onPause()
         findPreference<Preference>(getString(R.string.tileOverlaysKey))?.onPreferenceClickListener = null
         findPreference<Preference>(getString(R.string.onlineLayersKey))?.onPreferenceClickListener = null
      }

      private fun onFeeds(feeds: List<Feed>) {
         val screen = preferenceScreen
         val enabledFeeds = mapLayerPreferences.getEnabledFeeds(event?.id)
         for (feed in feeds) {
            val category = screen.getPreference(screen.preferenceCount - 1) as PreferenceCategory
            val feedPreference = mapLayerPreferences.mapFeedPreference(
               feed,
               requireActivity(),
               enabledFeeds.contains(feed.id)
            )
            feedPreference.onPreferenceClickListener =
               Preference.OnPreferenceClickListener {
                  onFeedClick(feed, feedPreference.isChecked)
                  true
               }
            category.addPreference(feedPreference)
         }
      }

      private fun onFeedClick(feed: Feed, on: Boolean) {
         val feeds = mapLayerPreferences.getEnabledFeeds(event?.id).toMutableSet()
         if (on) {
            feeds.add(feed.id)
         } else {
            feeds.remove(feed.id)
         }
         mapLayerPreferences.setEnabledFeeds(event?.id, feeds)
      }
   }

   public override fun onCreate(savedInstanceState: Bundle?) {
      super.onCreate(savedInstanceState)
      setContentView(R.layout.activity_map_preferences)

      val toolbar = findViewById<Toolbar>(R.id.map_preferences_toolbar)
      ViewCompat.setOnApplyWindowInsetsListener(toolbar) { v: View, windowInsets: WindowInsetsCompat ->
         val insets = windowInsets.getInsets(WindowInsetsCompat.Type.systemBars())
         v.setPadding(insets.left, 0, insets.right, 0)
         windowInsets
      }

      setSupportActionBar(toolbar)

      supportActionBar?.let {
         it.setDisplayHomeAsUpEnabled(true)
         it.setHomeAsUpIndicator(R.drawable.ic_arrow_back_white_24dp)
      }

      val fragContainer = findViewById<View>(R.id.map_preferences_fragment_container)
      ViewCompat.setOnApplyWindowInsetsListener(fragContainer) { v: View, windowInsets: WindowInsetsCompat ->
         val insets = windowInsets.getInsets(WindowInsetsCompat.Type.systemBars() or WindowInsetsCompat.Type.displayCutout())
         v.setPadding(insets.left, 0, insets.right, insets.bottom)
         windowInsets
      }

      supportFragmentManager.beginTransaction().replace(R.id.map_preferences_fragment_container, preference).commit()
   }

   companion object {
      const val TILE_OVERLAY_ACTIVITY = 100
      const val ONLINE_LAYERS_OVERLAY_ACTIVITY = 200
   }
}